src/librustc_mir/interpret/const_eval.rs

```diff
@@ -98,44 +98,20 @@ pub fn value_to_const_value<'a, 'tcx>(
     val: Value,
     ty: Ty<'tcx>,
 ) -> &'tcx ty::Const<'tcx> {
-    let layout = tcx.layout_of(ty::ParamEnv::reveal_all().and(ty)).unwrap();
-
-    if layout.is_zst() {
-        return ty::Const::from_const_value(
-            tcx,
-            ConstValue::ByVal(PrimVal::Undef),
```

Comment:
Is ZST always `ByVal(PrimVal::Undef)` in Miri?

Reply:
No. Accessing a zst field of a ByRef will return a ByRef

Reply:
Then this code should still be here.

Reply:
Why? Won't we simply get an empty allocation in the ByRef?

Reply:
Each Rust value should only map to one `ConstValue`. so a ZST type cannot be both `ByVal(PrimVal::Undef)` and `ByRef`.

Reply:
that would require us to spray is_zst checks everywhere. Miri already "violates" your assumption left and right.

The current code tries to reduce special cases, only ignoring zst memory reads in Memory, not earlier.

Reply:
Wait. wouldn't it actually make more sense to ~~try to~~ move to ByRef for zst everywhere? You can't ever have a zst value. They have no values (well one value, but we don't ever read/write it).